File: design-patterns/golang/builder/http_request_builder.go
package builder

import (
	"fmt"
	"strings"
)

// HttpRequestBuilder is responsible for building HttpRequest objects
type HttpRequestBuilder struct {
	url         string            // Required
	method      string            // Optional, default GET
	headers     map[string]string // Optional
	queryParams map[string]string // Optional
	body        string            // Optional
	timeout     int               // Optional, default 30s
}

// NewHttpRequestBuilder creates a new HttpRequestBuilder with the required URL
func NewHttpRequestBuilder(url string) (*HttpRequestBuilder, error) {
	if strings.TrimSpace(url) == "" {
		return nil, fmt.Errorf("URL cannot be null or empty")
	}

	return &HttpRequestBuilder{
		url:         url,
		method:      "GET",
		headers:     make(map[string]string),
		queryParams: make(map[string]string),
		timeout:     30000, // 30 seconds default
	}, nil
}

// Method sets the HTTP method
func (b *HttpRequestBuilder) Method(method string) *HttpRequestBuilder {
	if strings.TrimSpace(method) == "" {
		b.method = "GET"
	} else {
		b.method = strings.ToUpper(method)
	}
	return b
}

// Header adds a header to the request
func (b *HttpRequestBuilder) Header(key, value string) *HttpRequestBuilder {
	if key != "" && value != "" {
		b.headers[key] = value
	}
	return b
}

// QueryParam adds a query parameter to the request
func (b *HttpRequestBuilder) QueryParam(key, value string) *HttpRequestBuilder {
	if key != "" && value != "" {
		b.queryParams[key] = value
	}
	return b
}

// Body sets the request body
func (b *HttpRequestBuilder) Body(body string) *HttpRequestBuilder {
	b.body = body
	return b
}

// Timeout sets the request timeout in milliseconds
func (b *HttpRequestBuilder) Timeout(timeoutMillis int) *HttpRequestBuilder {
	if timeoutMillis > 0 {
		b.timeout = timeoutMillis
	}
	return b
}

// Build creates and returns a new HttpRequest
func (b *HttpRequestBuilder) Build() *HttpRequest {
	// Validate the request
	if (b.method == "POST" || b.method == "PUT") && b.body == "" {
		fmt.Printf("Warning: Building %s request without a body for URL: %s\n", b.method, b.url)
	}

	return NewHttpRequest(b)
}

File: design-patterns/cpp/builder/http_request_builder.cpp
#include "http_request_builder.h"
#include <iostream>
#include <algorithm>

HttpRequest::Builder::Builder(const std::string& url) : url(url) {
    if (url.empty()) {
        throw std::invalid_argument("URL cannot be empty.");
    }
}

HttpRequest::Builder& HttpRequest::Builder::method(const std::string& m) {
    method = m.empty() ? "GET" : m;
    std::transform(method.begin(), method.end(), method.begin(), ::toupper);
    return *this;
}

HttpRequest::Builder& HttpRequest::Builder::header(const std::string& key, const std::string& value) {
    if (!key.empty() && !value.empty()) headers[key] = value;
    return *this;
}

HttpRequest::Builder& HttpRequest::Builder::queryParam(const std::string& key, const std::string& value) {
    if (!key.empty() && !value.empty()) queryParams[key] = value;
    return *this;
}

HttpRequest::Builder& HttpRequest::Builder::body(const std::string& b) {
    body = b;
    return *this;
}

HttpRequest::Builder& HttpRequest::Builder::timeout(int timeoutMillis) {
    if (timeoutMillis > 0) timeout = timeoutMillis;
    return *this;
}

HttpRequest HttpRequest::Builder::build() {
    if ((method == "POST" || method == "PUT") && body.empty()) {
        std::cout << "Warning: Building " << method << " request without a body for URL: " << url << std::endl;
    }
    return HttpRequest(*this);
} 